src/utils/encoding/cid_de_cbor.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use cid::serde::BytesToCidVisitor;
+use cid::Cid;
+use core::fmt;
+use serde::de::{self, DeserializeSeed, SeqAccess, Visitor};
+use serde::Deserializer;
+use std::ops::{Deref, DerefMut};
+
+/// [`CidVec`] allows for efficient zero-copy de-serialization of `DAG_CBOR`-encoded nodes into a
+/// vector of [`Cid`].
+#[derive(Default)]
+pub struct CidVec(Vec<Cid>);
+
+impl Deref for CidVec {
+    type Target = Vec<Cid>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
```

Comment:
> Implementing Deref for smart pointers makes accessing the data behind them convenient, which is why they implement Deref. On the other hand, the rules regarding Deref and [DerefMut](https://doc.rust-lang.org/std/ops/trait.DerefMut.html) were designed specifically to accommodate smart pointers. Because of this, **Deref should only be implemented for smart pointers** to avoid confusion.

- https://doc.rust-lang.org/std/ops/trait.Deref.html (emphasis not mine)

Not a smart pointer

Reply:
I would have gone with `into_inner` here.

Reply:
Alright, `into_inner` it is then.

Reply:
> not using helpful abstractions in this case seems wasteful.

It's the wrong abstraction, and a commonly misused one at that, which is why it's called out in the docs

Reply:
> It's the wrong abstraction, and a commonly misused one at that, which is why it's called out in the docs

Sure is, though I'm still of the opinion that thin wrappers could benefit a lot from this. I would go further and say that the need for such wrappers to override certain behaviours calls for something magic/transparent to make the wrapper adhere to the same interface as whatever is wrapped. I wish there was a more eloquent way of implementing custom ser/de for types, e.g. the ability to have an implementation for a type alias.

Personal opinions aside - this has been fixed.